### DISCUSSION TITLE
Is it possible to track attributes not during assignment?

### DISCUSSION THREAD

**Discussion by seniorsolt** [gaogaotiantian/viztracer#459] [General]
Is it possible to track the attributes of objects not during assignment, but, for example, when an object is passed as an argument to a function? It would be handy to see something like in pycharm:
![2024-07-31_18-43-06](https://github.com/user-attachments/assets/fa0f1cf1-5c73-4940-ba5c-703267557d43)

**Comment by gaogaotiantian** [comment]
So a debugger is entirely different than a function-level tracer. Basically what you meant is to log when an attribute is read instead of write. The read operation is normally much more frequent and less interesting to log. People often care more about when the variable is changed. In what scenario do you want to log when an attribute is read?

**Reply by seniorsolt** [reply]
I actually use a debugger more often than a tracer or profiler to study unfamiliar programs or fix bugs. In this case, I like to read the trace like a book rather than step through it as in a classic debugger. I found the --log_func_args function and can indeed see the arguments within the current function, but if an argument is a complex object with many attributes, I don't see them. This means I need to find the previous place in the code where this object appears and where its attribute is being set.

Perhaps this option already exists, and I just don't see it.

For example:
Function \<predict\> takes \<config\> as an argument and it's an object of TrainConfig with lots of attributes.
def predict(
            self,
            model: StableDiffusionModel,
            batch: dict,
            config: TrainConfig,
            train_progress: TrainProgress,
            *,
            deterministic: bool = False,
    ) -> dict:
![2024-07-31_23-49-04](https://github.com/user-attachments/assets/6fa9284b-0cb9-42f5-b4a2-7a3b046962ca)

**Comment by gaogaotiantian** [comment]
Well, the way I can think of is to use `objprint` to print the data instead of `repr`. This feature itself is not easy to specify without source change.

**Reply by seniorsolt** [reply]
Got an idea, maybe I could record the attributes using sys.settrace and then add them to the final JSON

**Reply by gaogaotiantian** [reply]
You can't do `sys.settrace` together with `sys.setprofile`.

**Reply by seniorsolt** [reply]
why?

![2024-08-01_18-02-50](https://github.com/user-attachments/assets/8ba6fe12-753e-4b94-8303-0270edee3019)

also I think I could change void log_func_args from snaptrace.c to include attributes into repr

**Comment by gaogaotiantian** [comment]
Ah right I was thinking about nested `sys.settrace` and `sys.setprofile`. But no, `sys.settrace` will not be introduced in VizTracer due to its overhead. Also it's not easy to track that. Actually tracking an attribute is not trivial, you can check my project `watchpoints`.

Only include attributes in repr is not the ideal solution - what if your attributes have attributes? This is what I mentioned above - we can use `objprint` instead of repr to log the object. It takes some time to implement that - we need a new flag for users to specify it.

**Reply by seniorsolt** [reply]
I see. Okay. Objprint is a good solution. The only thing that bothers me is that the original objprint involves selective manual application to individual functions in the original code, for example, through a decorator. How do you plan to implement this in viztracer? Or will there be an option to apply objprint to all functions whose events are being tracked?

**Reply by gaogaotiantian** [reply]
No it does not. You can simply do `import objprint` then `objprint.op(obj)`. Or if you want the string, `objprint.objstr(obj)`.

**Reply by seniorsolt** [reply]
It's the same) It means that I need to choose what I want to track and write objprint.op(obj) manually. It’s not much different from regular logging or using print

**Reply by gaogaotiantian** [reply]
No, the user should be able to set a flag like `log_use_objprint=True` then `log_func_args` will use `objprint.op` instead of `repr` to log all the arguments.

**Reply by seniorsolt** [reply]
It would be great!

**Reply by gaogaotiantian** [reply]
I'm tracking this in #460 and I will implement it when I have some time. I'm a bit busy recently :)
